### PR TITLE
Stream Authorization in Programs

### PR DESCRIPTION
**Pull request by gokulavasan**
Support for Stream Authorization in Worker (for writing to streams), MR, Spark and Flows (latter three for reading).

Unit Tests for the all the above cases in the respective commits.

Build : http://builds.cask.co/browse/CDAP-DUT4559
ITN : http://builds.cask.co/browse/CDAP-ITM-35
JIRA : https://issues.cask.co/browse/CDAP-886

TODO: 
i) Should handle the stream authorization failure case for a continuously running program like Flowlet (since if they will keep logging failures) --> may be increase backoff or detect UnauthorizedException and handle it differently
ii) Add stream authorization in explore service
iii) If possible, add a check in Flow/MR/Spark before the flow begins so that we can fail fast if possible and not register usage/lineage etc.

### CODE REVIEW

**Comment by anwar6953** [comment]
This seems to handle user-written MapReduce jobs reading from streams, but what about explore-launched jobs that read from streams?

**Comment by gokulavasan** [comment]
@anwar6953 Thanks for bringing that up! I have added that item to the TODO list. It would be great to get a review on the current approach before I also adopt the same approach for stream authorization in Explore.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 1]
```diff
@@ -176,6 +189,17 @@ public StreamBatchWriter createBatchWriter(String stream, String contentType) th
...
   }
 
+  private HttpRequest.Builder addUserIdHeader(HttpRequest.Builder builder) throws IOException {
+    if (!kerberosEnabled) {
+      return builder;
+    }
+    return builder.addHeader(Constants.Security.Headers.USER_ID, getUserId());
+  }
+
+  private String getUserId() {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: need this method?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 2]
```diff
@@ -163,6 +173,9 @@ public StreamBatchWriter createBatchWriter(String stream, String contentType) th
     connection.setReadTimeout(15000);
     connection.setConnectTimeout(15000);
     connection.setRequestProperty(HttpHeaders.CONTENT_TYPE, contentType);
+    if (kerberosEnabled) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: how come we didn't need this till now? This seems unrelated to authorization, right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java [hunk 3]
```diff
@@ -75,16 +81,30 @@ public static String getWrappedMapper(Configuration conf) {
   public void run(Context context) throws IOException, InterruptedException {
     MapReduceClassLoader classLoader = MapReduceClassLoader.getFromConfiguration(context.getConfiguration());
     BasicMapReduceTaskContext basicMapReduceContext = classLoader.getTaskContextProvider().get(context);
+    AuthorizationEnforcer authorizationEnforcer = classLoader.getTaskContextProvider().getAuthorizationEnforcer();
+    AuthenticationContext authenticationContext = classLoader.getTaskContextProvider().getAuthenticationContext();
 
-    // this is a hook for periodic flushing of changes buffered by datasets (to avoid OOME)
-    WrappedMapper.Context flushingContext = createAutoFlushingContext(context, basicMapReduceContext);
+    EntityId entityId = null;
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java]: nit: move closer to where this is being used (line 96)

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java (hunk 2 above)
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: this should have been `authorizationEnabled`. And it should fetch the value of `security.authorization.enabled`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java [hunk 4]
```diff
@@ -156,6 +178,16 @@ public void run(Context context) throws IOException, InterruptedException {
 
       @Override
       public boolean nextKeyValue() throws IOException, InterruptedException {
+        // If entity id is not null, verify that the user has read access to the entity.
+        if (entityId != null) {
+          try {
+            authorizationEnforcer.enforce(entityId, authenticationContext.getPrincipal(), Action.READ);
+          } catch (Exception e) {
+            Throwables.propagateIfInstanceOf(e, IOException.class);
+            throw new IOException(e);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java]: so if there's an `UnauthorizedException` here, it will be propagated as an `IOException`? Hmmm, makes me think if we should make `UnauthorizedException` unchecked, and just throw as-is. Unrelated to this PR, but what do you think?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java [hunk 5]
```diff
@@ -62,12 +65,15 @@ protected void startUp() throws Exception {
...
+                                                                       metricsCollectionService,
+                                                                       authorizationEnforcementService).get();
       // All services should be started
       for (ListenableFuture<State> future : startFutures) {
         Preconditions.checkState(future.get() == State.RUNNING,
-                                 "Failed to start services: zkClient %s, kafkaClient %s, metricsCollection %s",
-                                 zkClientService.state(), kafkaClientService.state(), metricsCollectionService.state());
+                                 "Failed to start services: zkClient %s, kafkaClient %s, metricsCollection %s, " +
+                                   "authorizationEnforcementService %s",
+                                 zkClientService.state(), kafkaClientService.state(), metricsCollectionService.state(),
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java]: nit: unrelated to your change, but we probably don't need to do a `.state()` for any of these. Just print them as is, and Guava service will print it as `<Service Name> [Service State]`. That way you can remove the literals `zkClient`, `kafkaClient`, etc?

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamInputFormatTest.java [hunk 6]
```diff
@@ -297,6 +299,7 @@ public void testStreamRecordReader() throws Exception {
     // one from 0 - some offset and one from offset - Long.MAX_VALUE.
     Configuration conf = new Configuration();
     TaskAttemptContext context = new TaskAttemptContextImpl(conf, new TaskAttemptID());
+    StreamInputFormat.setStreamId(conf, DUMMY_ID);
```
Comment (bdmogal) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamInputFormatTest.java]: so users have to do this every time now? or is it only because this test tests `StreamInputFormat` in isolation?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java [hunk 7]
```diff
@@ -131,6 +133,16 @@ public static void setTimeRange(Configuration conf, long startTime, long endTime
   }
 
   /**
+   * Sets the stream id of the stream.
+   *
+   * @param conf The conf to modify.
+   * @param streamId {@link StreamId} id of the stream.
+   */
+  public static void setStreamId(Configuration conf, StreamId streamId) {
+    conf.set(STREAM_ID, streamId.toString());
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java]: should we gson serialize this instead of using `toString`? I remember we had some problems with deserializing using `EntityId.fromString` right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputSplit.java [hunk 8]
```diff
@@ -53,9 +56,10 @@ public StreamInputSplit() {
    * @param length Size of this split.
    * @param locations List of hosts containing this split.
    */
-  StreamInputSplit(Path path, @Nullable Path indexPath, long startTime, long endTime,
+  StreamInputSplit(StreamId streamId, Path path, @Nullable Path indexPath, long startTime, long endTime,
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputSplit.java]: `path` is not `@Nullable` anymore?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputSplit.java [hunk 9]
```diff
@@ -86,6 +97,7 @@ long getEndTime() {
   @Override
   public void write(DataOutput out) throws IOException {
     super.write(out);
+    WritableUtils.writeString(out, streamId.toString());
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputSplit.java]: should we use gson here too?

File: cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java [hunk 10]
```diff
@@ -423,7 +423,7 @@ public static void finish() throws Exception {
     authorizerInstantiator.close();
     streamCoordinatorClient.stopAndWait();
     metricsQueryService.stopAndWait();
-    metricsCollectionService.startAndWait();
```
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java:426]: nice :)

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 11]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
 
   @Test
   @Category(SlowTests.class)
+  public void testFlowStreamAuth() throws Exception {
+    createAuthNamespace();
+    Authorizer authorizer = getAuthorizer();
+    ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), StreamAuthApp.class);
+    final FlowManager flowManager = appManager.getFlowManager(StreamAuthApp.FLOW);
+    StreamId streamId = new StreamId(AUTH_NAMESPACE.getNamespace(), StreamAuthApp.STREAM);
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: can just do `AUTH_NAMESPACE.stream(StreamAuthApp.STREAM)`

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 12]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+      @Override
+      public Boolean call() throws Exception {
+        DataSetManager<KeyValueTable> kvTable = getDataset(AUTH_NAMESPACE.toId(), StreamAuthApp.KVTABLE);
+        return kvTable.get().read("Auth") != null;
+      }
+    }, 5, TimeUnit.SECONDS);
+    flowManager.stop();
+    flowManager.waitForFinish(5, TimeUnit.SECONDS);
+
+    // Now revoke write permission for Alice on that stream (revoke ALL and then grant everything other than READ)
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: you mean revoke read, right?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 13]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+
+  @Test
+  @Category(SlowTests.class)
+  public void testWorkerStreamAuth() throws Exception {
+    createAuthNamespace();
+    Authorizer authorizer = getAuthorizer();
+    ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), StreamAuthApp.class);
+    WorkerManager workerManager = appManager.getWorkerManager(StreamAuthApp.WORKER);
+    workerManager.start();
+    workerManager.waitForFinish(5, TimeUnit.SECONDS);
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:268]: how come we wait for a worker to finish without calling stop first?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 14]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+    try {
+      workerManager.stop();
+    } catch (Exception e) {
+      // workaround since we want worker job to be de-listed from the running processes to allow cleanup to happen
+      Assert.assertTrue(e.getCause() instanceof BadRequestException);
+    }
+    StreamId streamId = new StreamId(AUTH_NAMESPACE.getNamespace(), StreamAuthApp.STREAM);
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: `AUTH_NAMESPACE.stream(StreamAuthApp.STREAM)`

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 15]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+    StreamManager streamManager = getStreamManager(AUTH_NAMESPACE.toId(), StreamAuthApp.STREAM);
+    Assert.assertEquals(5, streamManager.getEvents(0, Long.MAX_VALUE, Integer.MAX_VALUE).size());
+
+    // Now revoke write permission for Alice on that stream (revoke ALL and then grant everything other than WRITE)
+    authorizer.revoke(streamId, ALICE, ImmutableSet.of(Action.ALL));
+    authorizer.grant(streamId, ALICE, ImmutableSet.of(Action.READ, Action.ADMIN, Action.EXECUTE));
+    workerManager.start();
+    workerManager.waitForFinish(5, TimeUnit.SECONDS);
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:283]: same question here

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 16]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+    try {
+      mrManager.stop();
+    } catch (Exception e) {
+      // workaround since we want mr job to be de-listed from the running processes to allow cleanup to happen
+      Assert.assertTrue(e.getCause() instanceof BadRequestException);
+    }
+    DataSetManager<KeyValueTable> kvManager = getDataset(AUTH_NAMESPACE.toId(), StreamAuthApp.KVTABLE);
+    KeyValueTable kvTable = kvManager.get();
+    byte[] value = kvTable.read("Hello");
+    Assert.assertArrayEquals(value, Bytes.toBytes("Hello"));
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: expected value should be the first argument?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 17]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+    kvTable.close();
+    // Since Alice had full permissions, she should be able to execute the MR job successfully
+    Tasks.waitFor(1, new Callable<Integer>() {
+      @Override
+      public Integer call() throws Exception {
+        return mrManager.getHistory(ProgramRunStatus.COMPLETED).size();
+      }
+    }, 5, TimeUnit.SECONDS);
+
+    ProgramId mrId = new ProgramId(AUTH_NAMESPACE.getNamespace(), StreamAuthApp.APP, ProgramType.MAPREDUCE,
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: `AUTH_NAMESPACE.app(StreamAuthApp.APP).mr(StreamAuthApp.MAPREDUCE)`

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 18]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+                                   StreamAuthApp.MAPREDUCE);
+    authorizer.grant(mrId.getNamespaceId(), BOB, ImmutableSet.of(Action.ALL));
+    ArtifactSummary artifactSummary = appManager.getInfo().getArtifact();
+
+    ArtifactId artifactId = new ArtifactId(AUTH_NAMESPACE.getNamespace(), artifactSummary.getName(),
+                                           artifactSummary.getVersion());
+    authorizer.grant(artifactId, BOB, ImmutableSet.of(Action.ALL));
+    authorizer.grant(mrId.getParent(), BOB, ImmutableSet.of(Action.ALL));
+    authorizer.grant(mrId, BOB, ImmutableSet.of(Action.ALL));
+    authorizer.grant(new StreamId(AUTH_NAMESPACE.getNamespace(), StreamAuthApp.STREAM), BOB,
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: ditto

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 19]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+                     ImmutableSet.of(Action.ADMIN));
+    authorizer.grant(new DatasetId(AUTH_NAMESPACE.getNamespace(), StreamAuthApp.KVTABLE), BOB,
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: ditto

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 20]
```diff
@@ -182,6 +194,196 @@ public void testNamespaces() throws Exception {
...
+      }
+    }, 5, TimeUnit.SECONDS);
+    kvManager = getDataset(AUTH_NAMESPACE.toId(), StreamAuthApp.KVTABLE);
+    kvTable = kvManager.get();
+    value = kvTable.read("World");
+    Assert.assertNull(value);
+    kvTable.close();
+
+    // Now grant Bob, READ access on the stream. MR job should execute successfully now.
+    authorizer.grant(new StreamId(AUTH_NAMESPACE.getNamespace(), StreamAuthApp.STREAM), BOB,
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: ditto

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/StreamAuthApp.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * App which copies data from a stream to KVTable using a MapReduce program.
```
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/StreamAuthApp.java]: nit: plus other stuff, right? make it generic ? "App to test authorization in programs"?

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamInputFormatTest.java (hunk 6 above)
Comment (gokulavasan) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamInputFormatTest.java]: users who? StreamInputFormat is not in api. Yes, it is because it tests StreamInputFormat in isolation.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputSplit.java (hunk 8 above)
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputSplit.java]: Path has never been Nullable. It is indexPath that is nullable.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java (hunk 7 above)
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java]: We had problems deserializing using EntityId.fromString?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 13 above)
Comment (gokulavasan) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:268]: worker here doesn't run forever, instead it sends 5 events to stream and stops.

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 15 above)
Comment (gokulavasan) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:283]: worker logic doesn't run forever.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java (hunk 7 above)
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java]: Oh I checked the code and it looks like the problem arises only for Datasets since we allow '.' in dataset IDs but we use '.' as a separator as well.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamInputFormat.java]: yeah https://issues.cask.co/browse/CDAP-5730. Better to not depend on the `toString`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java (hunk 4 above)
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java]: Yes, that would be good.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java [hunk 22]
```diff
@@ -245,6 +252,14 @@ public void flushOperations() throws Exception {
     }
   }
 
+  public AuthorizationEnforcer getAuthorizationEnforcer() {
+    return authorizationEnforcer;
+  }
+
+  public AuthenticationContext getAuthenticationContext() {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java:266]: since these are not overriden methods, would it be good to add a small javadoc (or even code comment) on why these need to be exposed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/stream/StreamInputFormatProvider.java [hunk 23]
```diff
@@ -14,14 +14,16 @@
  * the License.
  */
 
-package co.cask.cdap.data.stream;
+package co.cask.cdap.internal.app.runtime.batch.stream;
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/stream/StreamInputFormatProvider.java:17]: why do we need to move this? because it is a runtime class?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamInputFormat.java [hunk 24]
```diff
@@ -232,17 +234,20 @@ public static void inferDecoderClass(Configuration conf, Type vClass) {
                                          "StreamEventData if no decoder type is provided");
   }
 
+  public abstract AuthorizationEnforcer getAuthorizationEnforcer(TaskAttemptContext context);
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamInputFormat.java:241]: javadoc for these methods

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamRecordReader.java [hunk 25]
```diff
@@ -49,10 +58,14 @@
...
    * @param decoder The decoder to use for decoding stream events.
    */
-  StreamRecordReader(StreamEventDecoder<K, V> decoder) {
+  StreamRecordReader(StreamEventDecoder<K, V> decoder, AuthorizationEnforcer authorizationEnforcer,
+                     AuthenticationContext authenticationContext, StreamId streamId) {
     this.decoder = decoder;
     this.events = Lists.newArrayListWithCapacity(1);
     this.currentEntry = new StreamEventDecoder.DecodeResult<>();
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.principal = authenticationContext.getPrincipal();
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamRecordReader.java:68]: just checking: the record reader is created for every MR job, so its ok to cache the principal? Maybe add a note again

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamInputFormat.java [hunk 26]
```diff
@@ -234,8 +235,14 @@ public static void inferDecoderClass(Configuration conf, Type vClass) {
                                          "StreamEventData if no decoder type is provided");
   }
 
+  /**
+   * Return {@link AuthorizationEnforcer} to enforce authorization permissions.
+   */
   public abstract AuthorizationEnforcer getAuthorizationEnforcer(TaskAttemptContext context);
 
+  /**
+   * Return {@link AuthenticationContext} to fetch the {@link Principal}.
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamInputFormat.java]: nit: s/fetch/determine/?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java [hunk 27]
```diff
@@ -76,16 +76,14 @@ public void run(Context context) throws IOException, InterruptedException {
...
-    if (inputSplit instanceof TaggedInputSplit) {
-      TaggedInputSplit taggedInputSplit = (TaggedInputSplit) inputSplit;
-      basicMapReduceContext.setInputName(taggedInputSplit.getName());
-    }
-
     // this is a hook for periodic flushing of changes buffered by datasets (to avoid OOME)
     WrappedMapper.Context flushingContext = createAutoFlushingContext(context, basicMapReduceContext);
 
     basicMapReduceContext.setHadoopContext(flushingContext);
+    InputSplit inputSplit = context.getInputSplit();
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java:83]: curious - what was the issue here?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 28]
```diff
@@ -815,9 +815,9 @@ public boolean accept(String className, URL classUrl, URL classPathUrl) {
         classes.add(inputFormatClass);
 
         // If it is StreamInputFormat, also add the StreamEventCodec class as well.
-        if (AbstractStreamInputFormat.class.isAssignableFrom(inputFormatClass)) {
+        if (MapReduceStreamInputFormat.class.isAssignableFrom(inputFormatClass)) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:818]: checking: no similar revert for spark?

**Comment by bdmogal** [comment]
minor nits. Otherwise LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java (hunk 27 above)
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java:83]: Oh, there is no issue here. I was just reverting back to how the code was before since we moved the authorization from this class to StreamInputFormat.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 28 above)
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:818]: Hmm, no not required. Actually, I could have just left it as AbstractStreamInputFormat (which was actually added in this PR) but since it is used in MapReduceRuntimeService, I thought we could just safely use this class.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamRecordReader.java (hunk 25 above)
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamRecordReader.java:68]: Ya its ok to principal. I will add a note.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/stream/StreamInputFormatProvider.java (hunk 23 above)
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/stream/StreamInputFormatProvider.java:17]: Well I wasn't sure about a good place to put it (since .data package in app-fabric didn't sound right). Open to suggestions.

**Comment by gokulavasan** [comment]
@bdmogal Addressed comments. Let me know if it looks good to you. I can squash and merge since only javadocs/comments were updated.

**Comment by bdmogal** [comment]
LGTM

**Comment by gokulavasan** [comment]
Merging. Thanks for the review @bdmogal ! I will open JIRAs for the issues mentioned in the PR.
